src/main/java/de/hysky/skyblocker/skyblock/dungeon/puzzle/boulder/BoulderSolver.java

```diff
@@ -0,0 +1,209 @@
+     *
+     * @param initialStates The list of initial game states from which to start the search.
+     * @return A list of coordinates representing the shortest path to solve the puzzle,
+     *         or null if no solution is found within the maximum number of iterations.
+     */
+    public static List<int[]> aStarSolve(List<GameState> initialStates) {
+        Set<GameState> visited = new HashSet<>();
+        PriorityQueue<Pair<GameState, List<int[]>>> queue = new PriorityQueue<>(new AStarComparator());
+
+        for (GameState initialState : initialStates) {
+            queue.add(new Pair<>(initialState, new ArrayList<>()));
+        }
+
+        int maxIterations = 10000;
+        int iterations = 0;
+
+        while (!queue.isEmpty() && iterations < maxIterations) {
+            Pair<GameState, List<int[]>> pair = queue.poll();
+            GameState state = pair.first;
+            List<int[]> path = pair.second;
+
+            if (state.isSolved()) {
+                return path;
+            }
+
+            if (visited.contains(state)) {
+                continue;
+            }
+            visited.add(state);
+
+            int[] currentCoord = {state.playerX, state.playerY};
+            path.add(currentCoord);
+
+            for (int[] direction : new int[][]{{-1, 0}, {0, -1}, {0, 1}, {1, 0}}) {
+                GameState newState = new GameState(state.grid, state.playerX, state.playerY);
+                if (newState.movePlayer(direction[0], direction[1])) {
+                    queue.add(new Pair<>(newState, new ArrayList<>(path)));
+                }
+            }
+            iterations++;
+        }
+
+        return null;
+    }
+
+    /**
+     * A comparator used to compare game states based on their A* search cost.
+     * States with lower costs are prioritized for exploration.
+     */
+    private static class AStarComparator implements Comparator<Pair<GameState, List<int[]>>> {
+        /**
+         * Compares two pairs of game states and their associated paths based on their costs.
+         *
+         * @param a The first pair to compare.
+         * @param b The second pair to compare.
+         * @return A negative integer if a has a lower cost than b,
+         *         a positive integer if a has a higher cost than b,
+         *         or zero if both have the same cost.
+         */
+        @Override
+        public int compare(Pair<GameState, List<int[]>> a, Pair<GameState, List<int[]>> b) {
+            int costA = a.second.size() + a.first.heuristic();
+            int costB = b.second.size() + b.first.heuristic();
+            return Integer.compare(costA, costB);
+        }
+    }
+
+    /**
+     * Represents a pair of objects, such as a game state and its associated path.
+     *
+     * @param <T> The type of the first object.
+     * @param <U> The type of the second object.
+     */
+    private record Pair<T, U>(T first, U second) {}
```

Comment:
`it.unimi.dsi.fastutil.Pair` should be a drop in replacement (except constructors calls has to be changed to `Pair.of`).